site-builder/server/server.ts

```diff
@@ -0,0 +1,586 @@
+			} else {
+				// throw error
+				reject(err);
+			}
+		});
+
+		// Handle errors from archiver
+		archive.on("error", (err: Error) => {
+			console.error("Archive error:", err);
+			reject(err);
+		});
+
+		// Handle errors from the pass-through stream
+		passThrough.on("error", (err) => {
+			console.error("Stream error:", err);
+			reject(err);
+		});
+
+		// Start the upload process
+		upload
+			.done()
+			.then((result) => {
+				console.log(`Archive upload completed: ${fileName}`);
+				resolve(result.Location!);
+			})
+			.catch((err) => {
+				console.error("Upload error:", err);
+				reject(err);
+			});
+
+		// Append files from the directory to the archive
+		archive.directory(sourceDir, false);
+
+		// Finalize the archive - this is when data actually starts flowing
+		archive.finalize();
+	});
+};
+
+const verifySiteBuilderToken = async (authHeader: string, communitySlug: string) => {
+	if (!authHeader || !authHeader.startsWith("Bearer ")) {
+		throw new Error("Invalid or missing authorization token");
+	}
+
+	const client = initClient(siteApi, {
+		baseUrl: env.PUBPUB_URL,
+		baseHeaders: {
+			Authorization: authHeader,
+		},
+	});
+
+	const response = await client.auth.check.siteBuilder({
+		params: {
+			communitySlug: communitySlug,
+		},
+	});
+
+	if (response.status === 200) {
+		return;
+	}
+
+	if (response.status === 401) {
+		return {
+			status: 401,
+			body: {
+				success: false,
+				message: `${response.body.code}: ${response.body.reason}`,
+			},
+		} as const;
+	}
+
+	throw new Error(`UNKNOWN ERROR: ${response.body}`);
+};
+
+const router = tsr.router(siteBuilderApi, {
+	build: async ({ body, headers }, ctx) => {
+		console.log("Build request received");
+
+		try {
+			const authHeader = headers.authorization;
+			const authToken = authHeader.replace("Bearer ", "");
+			const communitySlug = body.communitySlug;
+
+			const tokenVerification = await verifySiteBuilderToken(authHeader, communitySlug);
+
+			if (tokenVerification) {
+				return tokenVerification;
+			}
+
+			const siteUrl = body.siteUrl;
+			const uploadToS3Folder = body.uploadToS3Folder || false;
+			const timestamp = Date.now();
+
+			const distDir = `./dist/${communitySlug}`;
+
+			console.log({ msg: `Building site`, distDir, siteUrl, authToken, communitySlug, env });
+
+			const buildSuccess = await buildAstroSite({
+				outDir: distDir,
+				site: siteUrl,
+				vite: {
+					define: {
+						"import.meta.env.COMMUNITY_SLUG": JSON.stringify(communitySlug),
+						"import.meta.env.AUTH_TOKEN": JSON.stringify(authToken),
+						"import.meta.env.PUBPUB_URL": JSON.stringify(env.PUBPUB_URL),
+					},
+				},
+			});
+
+			if (!buildSuccess) {
+				return {
+					status: 500,
+					body: { success: false, message: "Build failed" },
+				};
+			}
+
+			console.log("Build successful");
+
+			let uploadResult: string | undefined;
+			let s3FolderUrl: string | undefined;
+			let s3FolderPath: string | undefined;
+			let error: Error | undefined;
```

Comment:
i tried to make the builds slightly parallelizable across communities, and i think it works? i think `buildAstroSite` should spin up its own process and use diff env vars across them, but i have yet to test it. it builds the stuff inside of different directories at least, but i reuses the same directory across different builds for the same community. that could probably be better